### PR TITLE
fake backoff implementation for speeding up flakey test

### LINKED ISSUE #21739: unversioned client unit test TestBackoffLifecycle failed
I met this in #21349.
http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-test-unit-integration/15730/testReport/junit/

This test should have nothing to do with kubelet, so it seems to be a flake.

### PR DESCRIPTION
**Pull request by jayunit100**
I  put a quick patch together which Fixes #21739 .

This backoff adds a `Sleep` implementation to the clock, so that we can implement the same abstraction of sleeping during the requests `time.Sleep` using `clock.Sleep`.

### CODE REVIEW

**Comment by eparis** [comment]
```
_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/clock.go:148: undefined: fmt
```

**Comment by jayunit100** [comment]
Part of me is skeptical of the fake clock (even though I admit we definetely need it)....  So I  repushed another update which makes sure the elapsed time is also correct.  This way the test has a two stage verification, first of Calculated Backoff, then of elapsed backoff.  

- adds `Sleep` abstraction to BackoffMgr
- publicizes FakeBackoff
- implements fake clock in the request_test
- verifies backoff by checking both the calculated and the "real" backoff based on fake elapsed time.

**Comment by jayunit100** [comment]
tl;dr , tested and working now,  this definetely should fix the flake.

- adds `Sleep` abstraction to BackoffMgr
- adds `Sleep` functionality to the Clock interface w/ corresponding unit test.
- publicizes FakeBackoff
- implements fake clock in the request_test
- verifies backoff by checking both the calculated and the "real" backoff based on fake elapsed time.

File: pkg/client/unversioned/urlbackoff.go [hunk 1]
```diff
@@ -95,3 +99,7 @@ func (b *URLBackoff) UpdateBackoff(actualUrl *url.URL, err error, responseCode i
 func (b *URLBackoff) CalculateBackoff(actualUrl *url.URL) time.Duration {
 	return b.Backoff.Get(b.baseUrlKey(actualUrl))
 }
+
+func (b *URLBackoff) Sleep(d time.Duration) {
```
Comment (jayunit100) [inline pkg/client/unversioned/urlbackoff.go]: Sleep is implemented inside urlbackoff to decouple the request from any notion of a depedency on the clock interface.  This is important because I think we may want to swap out the clock later for an external clock impl (i.e. https://github.com/jonboulle/clockwork)

File: pkg/util/backoff.go [hunk 2]
```diff
@@ -36,6 +36,15 @@ type Backoff struct {
 	perItemBackoff  map[string]*backoffEntry
 }
 
+func NewFakeBackOff(initial, max time.Duration, tc *FakeClock) *Backoff {
```
Comment (eparis) [inline pkg/util/backoff.go]: Why did this get moved here? It is only used in _test.go....

File: pkg/client/unversioned/request_test.go [hunk 3]
```diff
@@ -767,17 +767,26 @@ func TestBackoffLifecycle(t *testing.T) {
 	// which are used in the server implementation returning StatusOK above.
 	seconds := []int{0, 1, 2, 4, 8, 0, 1, 2, 4, 0}
 	request := c.Verb("POST").Prefix("backofftest").Suffix("abc")
+	clock := util.FakeClock{}
 	request.backoffMgr = &URLBackoff{
-		Backoff: util.NewBackOff(
+		// Use a fake backoff here to avoid flakes and speed the test up.
+		Backoff: util.NewFakeBackOff(
```
Comment (jayunit100) [inline pkg/client/unversioned/request_test.go]: @eparis the FakeBackoff is useful in other places to .  basically anyone should be able to use it to avoid creating a flakey test that relies on *gasp* real time :)

File: pkg/util/clock.go [hunk 4]
```diff
@@ -137,6 +142,10 @@ func (f *FakeClock) HasWaiters() bool {
 	return len(f.waiters) > 0
 }
 
+func (f *FakeClock) Sleep(d time.Duration) {
+	f.Step(d)
```
Comment (lavalamp) [inline pkg/util/clock.go]: I think this is a bit confusing. Typically only the test code changes the clock setting. So I think the default expectation people would have is that Sleep would *block* until the test advances the clock far enough (you could implement it via After).

I guess I don't mind this too much but you should explain in a comment.

Comment (jayunit100) [inline pkg/util/clock.go]: Happy to comment it. But: maybe the tests shouldn't be advancing the clock manually, and instead, sleeping the way normal non test scenarios do...

I actually wrote this so tests don't need to couple their impl to the clock mocking scheme ..... But maybe I'm not thinking of all the scenarios
